pkg/common-controller/groupsnapshot_controller_helper.go

```diff
@@ -485,6 +493,204 @@ func (ctrl *csiSnapshotCommonController) syncUnreadyGroupSnapshot(groupSnapshot
+				groupSnapshotContent.Spec.Driver)
+		}
+
+		volumeSnapshotContentName := CreateSnapshotContentNameForVolumeGroupSnapshotContent(
+			string(groupSnapshotContent.UID), volumeHandle)
+
+		volumeSnapshotName := CreateSnapshotNameForVolumeGroupSnapshotContent(string(groupSnapshotContent.UID), volumeHandle)
+		volumeSnapshotNamespace := groupSnapshotContent.Spec.VolumeGroupSnapshotRef.Namespace
+
+		volumeHandleData[volumeHandle] = snapshotContentData{
+			snapshotContentName: volumeSnapshotContentName,
+			persistentVolume:    pv,
+		}
+
+		volumeSnapshotContent := &crdv1.VolumeSnapshotContent{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: volumeSnapshotContentName,
+			},
+			Spec: crdv1.VolumeSnapshotContentSpec{
+				VolumeSnapshotRef: v1.ObjectReference{
+					Kind:      "VolumeSnapshots",
+					Name:      volumeSnapshotName,
+					Namespace: volumeSnapshotNamespace,
+				},
+				DeletionPolicy: groupSnapshotContent.Spec.DeletionPolicy,
+				Driver:         groupSnapshotContent.Spec.Driver,
+				Source: crdv1.VolumeSnapshotContentSource{
+					SnapshotHandle: &snapshotHandle,
+				},
+			},
+			Status: &crdv1.VolumeSnapshotContentStatus{
+				VolumeGroupSnapshotHandle: &groupSnapshotContent.Name,
+			},
+		}
+
+		if pv != nil {
+			volumeSnapshotContent.Spec.SourceVolumeMode = pv.Spec.VolumeMode
+		}
+
+		if groupSnapshotSecret != nil {
+			klog.V(5).Infof("createSnapshotsForGroupSnapshotContent: set annotation [%s] on volume snapshot content [%s].", utils.AnnDeletionSecretRefName, volumeSnapshotContent.Name)
+			metav1.SetMetaDataAnnotation(&volumeSnapshotContent.ObjectMeta, utils.AnnDeletionSecretRefName, groupSnapshotSecret.Name)
+
+			klog.V(5).Infof("createSnapshotsForGroupSnapshotContent: set annotation [%s] on volume snapshot content [%s].", utils.AnnDeletionSecretRefNamespace, volumeSnapshotContent.Name)
+			metav1.SetMetaDataAnnotation(&volumeSnapshotContent.ObjectMeta, utils.AnnDeletionSecretRefNamespace, groupSnapshotSecret.Namespace)
+		}
+
+		label := make(map[string]string)
+		label["volumeGroupSnapshotName"] = groupSnapshotContent.Spec.VolumeGroupSnapshotRef.Name
+		volumeSnapshot := &crdv1.VolumeSnapshot{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:       volumeSnapshotName,
+				Namespace:  volumeSnapshotNamespace,
+				Labels:     label,
+				Finalizers: []string{utils.VolumeSnapshotInGroupFinalizer},
+			},
+			Spec: crdv1.VolumeSnapshotSpec{
+				Source: crdv1.VolumeSnapshotSource{
+					VolumeSnapshotContentName: &volumeSnapshotContentName,
+				},
+			},
+		}
+
+		_, err := ctrl.clientset.SnapshotV1().VolumeSnapshotContents().Create(context.TODO(), volumeSnapshotContent, metav1.CreateOptions{})
+		if err != nil {
+			return groupSnapshotContent, fmt.Errorf(
+				"createSnapshotsForGroupSnapshotContent: creating volumesnapshotcontent %w", err)
+		}
+
+		_, err = ctrl.clientset.SnapshotV1().VolumeSnapshots(volumeSnapshotNamespace).Create(context.TODO(), volumeSnapshot, metav1.CreateOptions{})
+		if err != nil {
```

Comment:
Picking a very random error here.

What happens when any Create / Patch calls here fail? To me it _seems_ that a next sync loop will pick _different_ object names (because they are based on the current time) and create *new* contents and snapshots from scratch. Leaving the old ones orphaned. That does not look good.

I think the `GroupSnapshotContent.Name` and `GroupSnapshot.Name` are unique enough (are they?) and all VolumeSnapshotContents could be based on its name (like `groupSnapshotContent.Name + "-" + pv.Name`). Similarly, for VolumeSnapshots. The names should be stable for a given source PV / volumeHandle and VolumeGroupSnapshot.

And then the code should ignore `AlreadyExists` errors.

Reply:
I agree with you, but I'm unaware of the underlying reasoning.

That's also why I changed the name of the original function `GetSnapshotContentNameForVolumeGroupSnapshotContent` to `CreateSnapshotContentNameForVolumeGroupSnapshotContent`.

Also, most bookkeeping could be simplified by an idempotent name.
I'm sure @xing-yang knows the context.

Reply:
That's a good point. This timestamp is even in the KEP.  This came from the beginning when we still have VolumeGroup in the design.  Though it does not look like we need it any more.  @leonardoce can you get rid of it for both VolumeSnapshot name and VolumeSnapshotContent name?

Reply:
Sure! Now, the names are consistent.